internal/client.go

```diff
@@ -228,63 +231,92 @@ func GetOrNewRocketMQClient(option ClientOptions, callbackCh chan interface{}) *
 func (c *rmqClient) Start() {
 	//ctx, cancel := context.WithCancel(context.Background())
 	//c.cancel = cancel
-	c.close = false
 	c.once.Do(func() {
-		// TODO fetchNameServerAddr
 		if !c.option.Credentials.IsEmpty() {
 			c.remoteClient.RegisterInterceptor(remote.ACLInterceptor(c.option.Credentials))
 		}
+		// TODO fetchNameServerAddr
 		go func() {}()
 
 		// schedule update route info
 		go func() {
 			// delay
+			ticker := time.NewTicker(_PullNameServerInterval)
+			defer ticker.Stop()
 			time.Sleep(50 * time.Millisecond)
-			for !c.close {
-				c.UpdateTopicRouteInfo()
-				time.Sleep(_PullNameServerInterval)
+			for {
+				select {
+				case <-ticker.C:
+					c.UpdateTopicRouteInfo()
+				case <-c.done:
+					rlog.Info("Stopping update topic route info", nil)
+					return
+				}
 			}
 		}()
 
 		// TODO cleanOfflineBroker & sendHeartbeatToAllBrokerWithLock
 		go func() {
-			for !c.close {
-				c.namesrvs.cleanOfflineBroker()
-				c.SendHeartbeatToAllBrokerWithLock()
-				time.Sleep(_HeartbeatBrokerInterval)
+			ticker := time.NewTicker(_HeartbeatBrokerInterval)
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ticker.C:
+					c.namesrvs.cleanOfflineBroker()
+					c.SendHeartbeatToAllBrokerWithLock()
+				case <-c.done:
+					rlog.Info("Stopping clean off line broker and heart beat", nil)
+					return
+				}
 			}
 		}()
 
 		// schedule persist offset
 		go func() {
-			//time.Sleep(10 * time.Second)
-			for !c.close {
-				c.consumerMap.Range(func(key, value interface{}) bool {
-					consumer := value.(InnerConsumer)
-					err := consumer.PersistConsumerOffset()
-					if err != nil {
-						rlog.Error("persist offset failed", map[string]interface{}{
-							rlog.LogKeyUnderlayError: err,
-						})
-					}
-					return true
-				})
-				time.Sleep(_PersistOffset)
+			ticker := time.NewTicker(_PersistOffset)
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ticker.C:
+					c.consumerMap.Range(func(key, value interface{}) bool {
+						consumer := value.(InnerConsumer)
+						err := consumer.PersistConsumerOffset()
+						if err != nil {
+							rlog.Error("persist offset failed", map[string]interface{}{
+								rlog.LogKeyUnderlayError: err,
+							})
+						}
+						return true
+					})
+				case <-c.done:
+					rlog.Info("Stopping persist offset", nil)
+					return
+				}
 			}
 		}()
 
 		go func() {
-			for !c.close {
-				c.RebalanceImmediately()
-				time.Sleep(_RebalanceInterval)
+			ticker := time.NewTicker(_RebalanceInterval)
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ticker.C:
+					c.RebalanceImmediately()
+				case <-c.done:
+					rlog.Info("Stopping do rebalance", nil)
+					return
+				}
 			}
 		}()
 	})
 }
 
 func (c *rmqClient) Shutdown() {
-	c.remoteClient.ShutDown()
-	c.close = true
+	c.shutdownOnce.Do(func() {
+		close(c.done)
+		c.close = true
```

Comment:
close filed maybe extra. why not only use c.done ?

Reply:
the close field used to judge client state, refer: https://github.com/apache/rocketmq-client-go/pull/303/files#diff-c9be6944b26f165a41c612366fa2b4e1R340